### PR TITLE
Structure Reindex Provider (relating to existing PR)

### PR DESCRIPTION
**Pull request by dantleech**
Changes relating to refactoring applying to existing PR #1981 

```
provider "doctrine_orm"

-- reindexing "11" instances of "Sulu\Bundle\ContactBundle\Entity\Contact"

 11/11 [============================] 100% Mem: 47,433,952b

-- reindexing "21" instances of "Sulu\Bundle\ContactBundle\Entity\Account"

 21/21 [============================] 100% Mem: 47,859,912b

-- reindexing "48" instances of "Sulu\Bundle\MediaBundle\Entity\FileVersionMeta"

 48/48 [============================] 100% Mem: 47,680,384b

provider "sulu_structure"

-- reindexing "373" instances of "Sulu\Bundle\ContentBundle\Document\PageDocument"

 373/373 [============================] 100% Mem: 85,070,424b

-- reindexing "1" instances of "Sulu\Bundle\ContentBundle\Document\HomeDocument"

 1/1 [============================] 100% Mem: 85,970,872b

-- reindexing "25" instances of "Sulu\Bundle\SnippetBundle\Document\SnippetDocument"

 25/25 [============================] 100% Mem: 85,385,104b

Index rebuild completed (36.48s 85,318,136b)
```

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getClassFqns()
+    {
+        $classFqns = array();
+        foreach ($this->metadataFactory->getAllMetadata() as $metadata) {
+            if (!$this->structureFactory->hasStructuresFor($metadata->getAlias())) {
+                continue;
+            }
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php]: Need to take into account Security and maybe some other things. check.

**Comment by dantleech** [comment]
Build failing for usual reason (passed on PHP7 however).

Please review, @chirimoya @danrot @wachterjohannes

File: src/Sulu/Bundle/ContentBundle/Tests/Unit/Search/ReIndex/StructureProviderTest.php [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+/*
+ * This file is part of Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Tests\Unit\Search\ReIndex;
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Tests/Unit/Search/ReIndex/StructureProviderTest.php]: Would also change the namespace here to `Reindex`

**Comment by danrot** [comment]
@dantleech Changelog missing.

File: src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+/*
+ * This file is part of Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Search\ReIndex;
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php]: Should also be `Reindex`

File: src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+use Massive\Bundle\SearchBundle\Search\ReIndex\LocalizedReIndexProviderInterface;
+use Sulu\Component\Content\Document\Behavior\SecurityBehavior;
+use Sulu\Component\Content\Metadata\Factory\StructureMetadataFactoryInterface;
+use Sulu\Component\DocumentManager\DocumentInspector;
+use Sulu\Component\DocumentManager\DocumentManagerInterface;
+use Sulu\Component\DocumentManager\MetadataFactoryInterface;
+
+class StructureProvider implements LocalizedReIndexProviderInterface
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Search/ReIndex/StructureProvider.php]: PHPDoc missing

**Comment by dantleech** [comment]
Cherry-picked into #1981

**Pull request closed without merging**